pkg/configstore/store.go

```diff
@@ -0,0 +1,172 @@
+package configstore
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+// Store holds config data
+type Store struct {
+	// If you care about sanity, never write here, just read
+	Data  map[string]interface{}
+	Dir   string
+	Hints map[string]string
+	sync.Mutex
+
+	ListFiles  func(string) ([]string, error)
+	ParseFile  func(string, interface{}) error
+	FileExists func(string) (bool, error)
+}
+
+func listFiles(dirname string) ([]string, error) {
+	files, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]string, len(files))
+	for idx, file := range files {
+		ret[idx] = file.Name()
+	}
+
+	return ret, nil
+}
+
+func parseFile(path string, value interface{}) error {
+	reader, err := os.Open(path)
+	defer reader.Close()
+	if err != nil {
+		return fmt.Errorf("Failed to open %s: %v", path, err)
+	}
+
+	return json.NewDecoder(reader).Decode(value)
+}
+
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// NewStore creates a new config store
+// `dir` is a directory from where data will be loaded
+// `hints` is a dictionary to find which file to load for
+// a given key
+func NewStore(dir string, hints map[string]string) *Store {
+	if hints == nil {
+		hints = map[string]string{}
+	}
+	return &Store{
+		Dir:        dir,
+		Hints:      hints,
+		ListFiles:  listFiles,
+		ParseFile:  parseFile,
+		FileExists: fileExists,
+	}
+}
+
+// Decode `path` contents using `json`, lock the store mutex, merge loaded data
+// into s.Data, unlock the mutex
+func (s *Store) loadPath(path string) error {
+	value := map[string]interface{}{}
+	err := s.ParseFile(path, value)
+	if err != nil {
+		return fmt.Errorf("Failed to parse %s: %v", path, err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for key, val := range value {
```

Comment:
Why do we need a mutex here, is there some concurrency going on?

Reply:
it's a library code that can be used in concurrent context, so it's better to protect racy sections

Reply:
wrapped map reads in a lock